cmd/snap-seccomp/main.go

```diff
@@ -554,8 +568,17 @@ func parseLine(line string, secFilter *seccomp.ScmpFilter) error {
 		return fmt.Errorf("too many arguments specified for syscall '%s' in line %q", tokens[0], line)
 	}
 
+	// allow the listed syscall but also support explicit denials as well by
+	// prefixing the line with a ~
+	action := seccomp.ActAllow
+
 	// fish out syscall
 	syscallName := tokens[0]
+	if strings.HasPrefix(syscallName, "~") {
```

Comment:
After some playing and poking I think this is okay short term but I think we should replace it ASAP in a followup. The issue is that it seems like libseccomp will "optimize" rules away. As it is designed as an allow-list, the behavior of this:
```
~ioctl - TIOCSTI
~ioctl - TIOCLINUX
ioctl
```
is counter intuitive - it will generate an allow rule for "ioctl" and discard the "~ioctl" entirely:
```
echo -e '~ioctl - TIOCSTI\nioctl\n' > seccomp-profile && SNAP_SECCOMP_DEBUG=1  ./snap-seccomp compile seccomp-profile /tmp/xxx
#
# pseudo filter code start
#
# filter for arch x86_64 (3221225534)
if ($arch == 3221225534)
  # filter for syscall "ioctl" (16) [priority: 65535]
  if ($syscall == 16)
    action ALLOW;
  # default action
  action ERRNO(1);
```
this only works as expected hen combined with another specific rule like:
```
~ioctl - TIOCSTI
~ioctl - TIOCLINUX
ioctl - !TIOCSTI
```
see
```
$ echo -e '~ioctl - TIOCSTI\n~ioctl - TIOCLINUX\nioctl - !TIOCSTI' > seccomp-profile && SNAP_SECCOMP_DEBUG=1  ./snap-seccomp compile seccomp-profile /tmp/xxx
#
# pseudo filter code start
#
# filter for arch x86_64 (3221225534)
if ($arch == 3221225534)
  # filter for syscall "ioctl" (16) [priority: 65532]
  if ($syscall == 16)
    if ($a1.hi32 == 0)
      if ($a1.lo32 == 21532)
        action ERRNO(13);
      if ($a1.lo32 == 21522)
        action ERRNO(13);
      else
        action ALLOW;
    else
      action ALLOW;
  # default action
  action ERRNO(1);

Reply:
I think we probably want instead of "~" a new: `ioctl - !TIOCSTI&&!TIOCLINUX`

Reply:
Unfortunately there is no way to express this in the libseccomp API https://libseccomp.readthedocs.io/en/latest/man/man3/seccomp_rule_add.3/ that I am aware of - we can only do something like 'compare argument 1 is equal to this value' or 'compare argument 1 is greater than this value' - we can't do higher level operations like booleans etc - so I don't think that is possible with libseccomp. If we were to write the BPF filter by hand we could add this but that would be a significant departure from the current implementation.